### ISSUE TITLE
Invalid device function dropout

### ISSUE THREAD

**Issue by ssakhavi** [wojzaremba/lstm#1]
I get the following error when running `main.lua`

stack traceback:
    [C]: in function 'copy'
    ...iavash/torch-distro/install/share/lua/5.1/nn/Dropout.lua:16: in function 'func'
    ...h/torch-distro/install/share/lua/5.1/nngraph/gmodule.lua:169: in function 'neteval'
    ...h/torch-distro/install/share/lua/5.1/nngraph/gmodule.lua:199: in function 'forward'
    main.lua:148: in function 'fp'
    main.lua:224: in function 'main'
    main.lua:259: in main chunk
    [C]: in function 'f'
    [string "local f = function() return dofile 'main.lua'..."]:1: in main chunk
    [C]: in function 'xpcall'
    ...iavash/torch-distro/install/share/lua/5.1/trepl/init.lua:583: in function 'repl'
    ...rch-distro/install/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
    [C]: at 0x00406170

**Comment by wojzaremba** [comment]
Probably you have an old version of torch. I have looked on few machines, and I do not experience this problem. 

I am suggesting to install facebook version of torch. It's very recent and stable. https://github.com/facebook/fblualib

**Comment by alexbw** [comment]
I'm seeing this as well. I've followed the instructions on a completely fresh machine for installing `fbcunn` exactly here: https://github.com/facebook/fbcunn/blob/master/INSTALL.md.

I then install nngraph like so:

```
luarocks install nngraph
```

I'm running the script like so:

```
th main.lua
```

The output is:

```
Loading ./data/ptb.train.txt, size of data = 929589
Loading ./data/ptb.valid.txt, size of data = 73760
Loading ./data/ptb.test.txt, size of data = 82430
Using 1-th gpu
Network parameters:
{
  layers : 2
  lr : 1
  max_max_epoch : 13
  max_grad_norm : 5
  max_epoch : 4
  init_weight : 0.1
  decay : 2
  vocab_size : 10000
  seq_length : 20
  batch_size : 20
  dropout : 0
  rnn_size : 200
}
Creating a RNN LSTM network.
Starting training.
/root/torch/install/bin/luajit: /root/torch/install/share/lua/5.1/nn/Dropout.lua:16: invalid device function
stack traceback:
    [C]: in function 'copy'
    /root/torch/install/share/lua/5.1/nn/Dropout.lua:16: in function 'func'
    /root/torch/install/share/lua/5.1/nngraph/gmodule.lua:169: in function 'neteval'
    /root/torch/install/share/lua/5.1/nngraph/gmodule.lua:199: in function 'forward'
    main.lua:148: in function 'fp'
    main.lua:224: in function 'main'
    main.lua:259: in main chunk
    [C]: in function 'dofile'
    /root/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
    [C]: at 0x00406170
```

(pardon the use of root user, I just got the fbcunn install working on root, and need to transfer it to a non-superuser).

**Comment by wojzaremba** [comment]
I will solve this issue. However, I am not experiencing this bug.
1. Could you check if everything works when you remove nn.Dropout layer. It's enough to modify line 85 to:
local dropped        = i[layer_idx - 1]
2. Could you turn on debuger or error (i.e. export LUA_DEBUG_ON_ERROR=1 ), and print out some more information about what is going on in line 16 in Dropout.

**Comment by soumith** [comment]
oh, i think i might know the issue. What graphics card are you guys running this on?

**Comment by ssakhavi** [comment]
My graphic card is NVIDIA Geforce GTX 770M on a gaming laptop.

On Sat, Feb 21, 2015 at 11:13 AM, Soumith Chintala <notifications@github.com

**Comment by soumith** [comment]
that's it. fbcunn is built for CUDA capability 3.5 or higher ONLY. the cards that you are running this repo are CUDA capable 3.0 (or less)

**Comment by soumith** [comment]
let's see if there can be a work-around. if cuda-capability 3.5 then use fbcunn's LookupTableGPU, if not use a slower version, i.e. nn's LookupTable

**Comment by ssakhavi** [comment]
Yep ...mine is 3.0. Thanks

**Comment by alexbw** [comment]
I'm using the g2.2xlarge, unsure of the CUDA compute capability but will
check soon.

EDIT: Yes, the g2.2xlarge has a GK104-based graphics card, which has CUDA compute capability at 3.0. The GK110 is the first chip in that line to support 3.5. 

This error message is a bit obscure, in which project should I open an issue about explicitly alerting users that certain functionality is unavailable due to the CUDA compute capability of their card? Should I use `fbcunn`, or `cunn`?

**Comment by alexbw** [comment]
And yes, changing LookupTableGPU to LookupTable worked! Thanks @soumith and @wojzaremba!

Curious what performance is with LookupTableGPU on a modern card. I'm assuming `wps` is the overall average words-per-second. A g2.2xlarge instance is producing `wps = 3495`.

**Comment by soumith** [comment]
The default config took me 29 mins in all.

**Comment by soumith** [comment]
On gtx980
